core/src/main/scala/caliban/validation/Validator.scala

```diff
@@ -1178,13 +1166,20 @@ object Validator {
         validateInputValueDirectives(field.allArgs, fieldErrorContext)
     }
 
-    ZPure.foreachDiscard(types) { t =>
+    validateAll(types) { t =>
       lazy val typeErrorContext = s"Type '${t.name.getOrElse("")}'"
       for {
         _ <- validateDirectives(t.directives, typeErrorContext)
         _ <- validateInputValueDirectives(t.allInputFields, typeErrorContext)
-        _ <- ZPure.foreachDiscard(t.allFields)(validateFieldDirectives(_, typeErrorContext))
+        _ <- validateAll(t.allFields)(validateFieldDirectives(_, typeErrorContext))
       } yield ()
     }
   }
+
+  // Pure's implementation doesn't check if the Iterable is empty, and this is causing some performance degradation
+  private def validateAll[R, A, B](
```

Comment:
How about adding this to `ZPure` so we ultimately don't need this?

Reply:
It's a bit tricky. The default implementation of `isEmpty` on an Iterable is to create an Iterator and call `.hasNext()` on it. In cases that the collection is non-empty, this can cause a big allocation. In our case, we're only using it with Lists and Maps, which override the default `.isEmpty` implementation.

Having said that, I don't know of any collections that don't override the isEmpty implementation so it might not be an issue 🤔

Reply:
Probably the lazy collections like Scala.Stream